### PR TITLE
Noc routing

### PR DESCRIPTION
**Pull request by Srivat97**
Added a number of routing algorithms that can be used to route traffic flows within the NoC during placement.

#### Description

- Added a base general routing class called NocRouting
- Added a routing class called XYRouting that runs the directional oriented routing algorithm
- Added a routing class called BFSRouting that runs breadth first search to find a route
- Added unit tests to verify the routing functions
- Added an extended functionality to the NoC model so that now links can be removed

#### Related Issue

N/A

#### Motivation and Context

We need to be able to evaluate whether placement of noc router blocks on the FPGA. We do this by determining and minimizing the aggregate bandwidth and latency's of traffic flows within the NoC. To determine the previous metrics we need to first find routes within the NoC that traffic flows can use. The purpose of this PR is provide routing algorithms that can be used to route traffic flows within the NoC. 

#### Types of changes

- [ ] Bug fix (change which fixes an issue)
- [x] New feature (change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Checklist:

- [x] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed

### CODE REVIEW

File: vpr/src/noc/xy_routing.h [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+    RouteDirection get_direction_to_travel(int sink_router_x_position, int sink_router_y_position, int curr_router_x_position, int curr_router_y_position);
+
+    /**
+     * @brief Given the direction to travel next, this function determines
+     * the outgoing link that should be used to travel in the inteded direction.
+     * Each router may have any number of outgoing links and each link is not
+     * guaranteed to point in the intended direction, So this function makes
+     * sure that the link chosen points in the intended direction.
+     * 
+     * @param curr_router_id The unique identifier of the current router that
```
Comment (vaughnbetz) [inline vpr/src/noc/xy_routing.h:148]: physical router (for id)?

File: vpr/src/noc/xy_routing.h [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+     * @param next_step_direction The direction to travel next
+     * @param flow_route Stores the path as a series of NoC links found by 
+     * a NoC routing algorithm between two routers in a traffic flow. The 
+     * NoC link found to travel next will be added to this path.
+     * @param visited_routers Keeps track of which routers have been reached
+     * already while traversing the NoC.
+     * @param noc_model A model of the NoC. This is used to traverse the
+     * NoC and find a route between the two routers.
+     * @return true A suitable link was found that we can traverse next
+     * @return false N suitable links were found that could be traversed
```
Comment (vaughnbetz) [inline vpr/src/noc/xy_routing.h:163]: No suitable link was found

Comment (vaughnbetz) [reply vpr/src/noc/xy_routing.h:163]: (N is typo): No

File: vpr/src/noc/xy_routing.cpp [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+     * */
+    while (curr_router_id != sink_router_id) {
+        // store the router that is currently visited at each iteration of the algorithm
+        const NocRouter curr_router = noc_model.get_single_noc_router(curr_router_id);
+
+        // get the position of the current router
+        int curr_router_x_position = curr_router.get_router_grid_position_x();
+        int curr_router_y_position = curr_router.get_router_grid_position_y();
+
+        // determine how we should path next
```
Comment (vaughnbetz) [inline vpr/src/noc/xy_routing.cpp:53]: "how we should path next" --> find direction to travel in next 
(or cut this comment, as pretty clear func name.

File: vpr/src/noc/xy_routing.cpp [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    if (curr_router_x_position > sink_router_x_position) {
+        direction_to_travel = RouteDirection::LEFT;
+    } else if (curr_router_x_position < sink_router_x_position) {
+        direction_to_travel = RouteDirection::RIGHT;
+    } else if (curr_router_y_position < sink_router_y_position) {
+        direction_to_travel = RouteDirection::UP;
+    } else if (curr_router_y_position > sink_router_y_position) {
+        direction_to_travel = RouteDirection::DOWN;
+    }
+
```
Comment (vaughnbetz) [inline vpr/src/noc/xy_routing.cpp:90]: Add last case or an assert (src and sink not at same location).

File: vpr/src/noc/noc_storage.h [hunk 5]
```diff
@@ -271,6 +271,26 @@ class NocStorage {
     void set_noc_router_latency(double router_latency);
 
     // general utiliy functions
+    /**
+     * @brief The link is indentified by going through the outoging links of the
```
Comment (vaughnbetz) [inline vpr/src/noc/noc_storage.h:275]: Start with what you do (removes a link between two specified physical routers).

Comment (vaughnbetz) [reply vpr/src/noc/noc_storage.h:275]: Update the link removal code to also change the data for that link (in the link vector) to invalid values, e.g. source = sink = RouterID::Invalid()

File: vpr/src/noc/noc_storage.cpp [hunk 6]
```diff
@@ -93,6 +93,46 @@ void NocStorage::set_noc_router_latency(double router_latency) {
...
+            }
+
+            outgoing_link_index++;
+        }
+    }
+
+    // if a link was not removed then throw warning message
+    if (!link_removed_status) {
+        VTR_LOG_WARN("No link could be found that has a source router with id: '%d' and sink router with id:'%d'.\n", (size_t)src_router_id, (size_t)sink_router_id);
+    }
```
Comment (vaughnbetz) [inline vpr/src/noc/noc_storage.cpp:131]: Also set data for this link to invalid (e.g. source = sink = RouterID::Invalid() ).

File: vpr/src/noc/bfs_routing.h [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * @brief Finds a route that goes from the starting router in a 
+     * traffic flow to the destination router. Uses the breadth first search
+     * algorithm to determine the route. A route consists of a series
+     * of links that should be traversed when travelling between two routers
+     * within the NoC.
+     * 
+     * @param src_router_id The source router of a traffic flow. Identifies 
+     * the starting point of the route within the NoC. This is represents
+     * a unique identifier of the source router.
```
Comment (vaughnbetz) [inline vpr/src/noc/bfs_routing.h:41]: Same comment as in the base class; mention physical routers and fix typos (if it has the same typos).

File: vpr/src/noc/bfs_routing.h [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+     * The function will clear any
+     * previously stored path and re-insert the new found path between the
+     * two routers. 
+     * @param noc_model A model of the NoC. This is used to traverse the
+     * NoC and find a route between the two routers.
+     * @param router_parent_link Contains the parent link associated to each
+     * router in the NoC (parent link is the link used to visit the router during
+     * the BFS routing algorithm).
+     */
+    void generate_route(NocRouterId sink_router_id, std::vector<NocLinkId>& flow_route, const NocStorage& noc_model, const std::unordered_map<NocRouterId, NocLinkId>& router_parent_link);
```
Comment (vaughnbetz) [inline vpr/src/noc/bfs_routing.h:77]: Seems like making these private would be good.

File: vpr/src/noc/bfs_routing.cpp [hunk 9]
```diff
@@ -0,0 +1,122 @@
...
+
+        // go through the outgoing links of the current router and process the sink routers connected to these links
+        for (auto link : outgoing_links) {
+            // get the router connected to the current outgoing link
+            NocRouterId connected_router = noc_model.get_single_noc_link(link).get_sink_router();
+
+            // check if we already explored the router we are visiting
+            if (visited_routers.find(connected_router) == visited_routers.end()) {
+                // if we are here then this is the first time visiting this router //
+                // mark this router as visited and add sned the router to be processed
```
Comment (vaughnbetz) [inline vpr/src/noc/bfs_routing.cpp:62]: sned -> send

**Review by vaughnbetz — CHANGES_REQUESTED** [review]
Looks good, but some changes requested.

File: vpr/src/noc/bfs_routing.h (hunk 8 above)
Comment (Srivat97) [reply vpr/src/noc/bfs_routing.h:77]: done

File: vpr/src/noc/noc_storage.cpp (hunk 6 above)
Comment (Srivat97) [reply vpr/src/noc/noc_storage.cpp:131]: done

File: vpr/src/noc/noc_routing.h [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+     * found, it can be accessed through this variable. 
+     * 
+     * A route consists of a series of links that should be traversed when
+     * beginning at a start router to reach the destination router. 
+     * 
+     * The derived classes that implement a NoC routing algorithm should
+     * all store the found route in this variable. Then external functions
+     * should access the found route through this variable.
+     */
+    std::vector<NocLinkId> routed_path;
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/noc_routing.h:48]: Done

File: vpr/src/noc/noc_routing.h [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+/**
+ * @file NocRouting.h
+ * @brief This file defines the NocRouting class, which handles the
+ * routing between routers within the NoC. It describes the routing
+ * algorithm for the NoC.
+ * 
+ * Overview
+ * ========
+ * The NocRouting class is an abstract class. It is intended to be used as
+ * a base class and should not be used on its own. The NocRouting
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/noc_routing.h:13]: done

File: vpr/src/noc/noc_routing.h [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+     * is that this function should be overridden in the derived classes
+     * to implement the routing algorithm.
+     * 
+     * @param src_router_id The source router of a traffic flow. Identifies 
+     * the starting point of the route within the NoC. This is represents
+     * a unique identifier of the source router.
+     * @param sink_router_id The destination router of a traffic flow.
+     * Identifies the ending point of the route within the NoC.This is 
+     * represents a unique identifier of the destination router.
+     * @param flow_route Stores the path as a series of NoC links found by
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/noc_routing.h:69]: kept this and added the recommended comment

File: vpr/src/noc/noc_routing.h [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+     * of links that should be traversed when travelling between two routers
+     * within the NoC. Derived classes will
+     * primarily differ by the routing algorithm they use. The expectation
+     * is that this function should be overridden in the derived classes
+     * to implement the routing algorithm.
+     * 
+     * @param src_router_id The source router of a traffic flow. Identifies 
+     * the starting point of the route within the NoC. This is represents
+     * a unique identifier of the source router.
+     * @param sink_router_id The destination router of a traffic flow.
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/noc_routing.h:66]: done

File: vpr/src/noc/noc_routing.h [hunk 14]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * @brief Finds a route that goes from the starting router in a 
+     * traffic flow to the destination router. A route consists of a series
+     * of links that should be traversed when travelling between two routers
+     * within the NoC. Derived classes will
+     * primarily differ by the routing algorithm they use. The expectation
+     * is that this function should be overridden in the derived classes
+     * to implement the routing algorithm.
+     * 
+     * @param src_router_id The source router of a traffic flow. Identifies
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/noc_routing.h:63]: done

File: vpr/src/noc/noc_routing.h [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+     * A route consists of a series of links that should be traversed when
+     * beginning at a start router to reach the destination router. 
+     * 
+     * The derived classes that implement a NoC routing algorithm should
+     * all store the found route in this variable. Then external functions
+     * should access the found route through this variable.
+     */
+    std::vector<NocLinkId> routed_path;
+
+    // protected pure virtual functions that should be implemendted in derived classes.
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/noc_routing.h:50]: done

File: vpr/src/noc/noc_routing.h [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+ * a base class and should not be used on its own. The NocRouting
+ * class describes the structure of what a routing algorithm class should
+ * look like. 
+ * 
+ * Usage
+ * -----
+ * When a new routing algorithm for the NoC is needed, a new class should be
+ * made that inherits this class. Then two things need to be done:
+ *  - The routing algorithm should be implemented inside the route_flow
+ *    function and should match the prototype dclared below
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/noc_routing.h:22]: done

File: vpr/src/noc/noc_routing.h [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+ *  - The found route should be stored inside the member variable defined below
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/noc_routing.h:23]: done

File: vpr/src/noc/noc_routing.h [hunk 18]
```diff
@@ -0,0 +1,80 @@
+#ifndef VTR_NOCROUTING_H
+#define VTR_NOCROUTING_H
+
+/**
+ * @file NocRouting.h
+ * @brief This file defines the NocRouting class, which handles the
+ * routing between routers within the NoC. It describes the routing
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/noc_routing.h:7]: done

File: vpr/src/noc/bfs_routing.cpp (hunk 9 above)
Comment (Srivat97) [reply vpr/src/noc/bfs_routing.cpp:62]: fixed

File: vpr/src/noc/bfs_routing.h (hunk 7 above)
Comment (Srivat97) [reply vpr/src/noc/bfs_routing.h:41]: fixed

File: vpr/src/noc/noc_storage.h (hunk 5 above)
Comment (Srivat97) [reply vpr/src/noc/noc_storage.h:275]: done

File: vpr/src/noc/xy_routing.cpp (hunk 3 above)
Comment (Srivat97) [reply vpr/src/noc/xy_routing.cpp:53]: comment removed

File: vpr/src/noc/xy_routing.h (hunk 2 above)
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:163]: fixed

File: vpr/src/noc/xy_routing.h [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+     * router that is currently being visited on the FPGA
+     * @param curr_router_y_position The vertical grid position of the router
+     * that is currently being visited on the FPGA
+     * @return RouteDirection The direction to travel next
+     */
+    RouteDirection get_direction_to_travel(int sink_router_x_position, int sink_router_y_position, int curr_router_x_position, int curr_router_y_position);
+
+    /**
+     * @brief Given the direction to travel next, this function determines
+     * the outgoing link that should be used to travel in the inteded direction.
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:143]: fixed

File: vpr/src/noc/xy_routing.h [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+
+    /**
+     * @brief Finds a route that goes from the starting router in a 
+     * traffic flow to the destination router. Uses the XY-routing
+     * algorithm to determine the route. A route consists of a series
+     * of links that should be traversed when travelling between two routers
+     * within the NoC.
+     * 
+     * @param src_router_id The source router of a traffic flow. Identifies 
+     * the starting point of the route within the NoC. This is represents
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:110]: done

File: vpr/src/noc/xy_routing.h [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+ * otherwise in moves in the direction to the destination router (once again
+ * the movement is done in the X-axis). 
+ * 
+ * Once horizontally aligned (current router in the path
+ * has the same X-coordinate as the destination) the algorithm
+ * checks to see whether the y-axis coordinates match between the destination 
+ * router and the current router in the path (checking for vertical alignment).
+ * Similiar to the x-axis movement, the algorithm moves in the Y-axis towards
+ * the destination router. Once again, at each router in the path the algorithm
+ * checks for vertical alignment, if not aligned it then moves in the y-axis
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:34]: done

File: vpr/src/noc/xy_routing.h [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+ * determined by the XY-Routing algorithm is shown as "<++++".
+ * 
+ * Note that this routing algorithm in inherently deadlock free.
+ * 
+ * Usage
+ * -----
+ * It is recommmended to use this algorithm when the NoC topology is of type
+ * Mesh. This algorithm will work for other types of toplogies but the
+ * directional nature of the algorithm make it ideal for mesh topologies. If
+ * the algorithm fails to find a router then an error is thrown.
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:80]: done

File: vpr/src/noc/xy_routing.h [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+ * and the router marked with the '$' character is the destination. The path
+ * determined by the XY-Routing algorithm is shown as "<++++".
+ * 
+ * Note that this routing algorithm in inherently deadlock free.
+ * 
+ * Usage
+ * -----
+ * It is recommmended to use this algorithm when the NoC topology is of type
+ * Mesh. This algorithm will work for other types of toplogies but the
+ * directional nature of the algorithm make it ideal for mesh topologies. If
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:79]: done

File: vpr/src/noc/xy_routing.h [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+ * The algorithm works by first travelling in the X-direction and then
+ * in the Y-direction. 
+ * 
+ * First the algorithm compares the source router and the 
+ * destination router, checking the corrdinates in the X-axis. If the
+ * corrdinates are not the same (so not horizontall aligned), then the
+ * algorithm moves in the direction towards the destination router in the 
+ * X-axis. For each router in the path, the algorithm checks again to see
+ * whether it is horizontall aligned with the destination router,
+ * otherwise in moves in the direction to the destination router (once again
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:25]: done

File: vpr/src/noc/xy_routing.h [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+ * --------------------
+ * The algorithm works by first travelling in the X-direction and then
+ * in the Y-direction. 
+ * 
+ * First the algorithm compares the source router and the 
+ * destination router, checking the corrdinates in the X-axis. If the
+ * corrdinates are not the same (so not horizontall aligned), then the
+ * algorithm moves in the direction towards the destination router in the 
+ * X-axis. For each router in the path, the algorithm checks again to see
+ * whether it is horizontall aligned with the destination router,
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:24]: done

File: vpr/src/noc/xy_routing.h [hunk 26]
```diff
@@ -0,0 +1,168 @@
...
+ * NoC. This class is based on the XY routing algorithm.
+ * 
+ * XY Routing Algorithm
+ * --------------------
+ * The algorithm works by first travelling in the X-direction and then
+ * in the Y-direction. 
+ * 
+ * First the algorithm compares the source router and the 
+ * destination router, checking the corrdinates in the X-axis. If the
+ * corrdinates are not the same (so not horizontall aligned), then the
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:21]: done

File: vpr/src/noc/xy_routing.h [hunk 27]
```diff
@@ -0,0 +1,168 @@
...
+#define VTR_XYROUTING_H
+
+/**
+ * @file 
+ * @brief This file defines the XYRouting class, which represents a directional
+ * oriented routing algorithm.
+ * 
+ * Overview
+ * ========
+ * The XYRouting class performs routing between routers in the
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:11]: done

File: vpr/src/noc/xy_routing.h [hunk 28]
```diff
@@ -0,0 +1,168 @@
+#ifndef VTR_XYROUTING_H
+#define VTR_XYROUTING_H
+
+/**
+ * @file 
+ * @brief This file defines the XYRouting class, which represents a directional
+ * oriented routing algorithm.
```
[the remark this answers is not in the document]
Comment (Srivat97) [reply vpr/src/noc/xy_routing.h:7]: done

File: vpr/src/noc/noc_routing_algorithm_creator.h [hunk 29]
```diff
@@ -0,0 +1,69 @@
+#ifndef NOC_ROUTING_ALGORITHM_CREATOR
+#define NOC_ROUTING_ALGORITHM_CREATOR
+
+/**
+ * @file om)
```
Comment (vaughnbetz) [inline vpr/src/noc/noc_routing_algorithm_creator.h:5]: om?

File: vpr/src/noc/noc_routing_algorithm_creator.h [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+ * the routing algorithm that will be used to route packets within the NoC.
+ * 
+ * Overview
+ * ========
+ * There are a number of different available NoC routing algorithms. The purpose of
+ * this class is to construct the appropriate routing algorithm based on the user
+ * specification in the command line. The user identifies a specific routing algorithm
+ * in the command line by providing a string (which is the name of routing algorithm).
+ * Then, the appripriate routing algorithm is created here based on the provided
+ * string.
```
Comment (vaughnbetz) [inline vpr/src/noc/noc_routing_algorithm_creator.h:16]: Might use the word factory here somewhere.

**Review by vaughnbetz — APPROVED** [review]
Minor commenting suggestions.

**Pull request merged**